src/compiler/checker.ts

```diff
@@ -9454,32 +9454,30 @@ namespace ts {
                         }
                     }
                 }
+                else if (isGenericMappedType(target) && !isGenericMappedType(source) && getConstraintTypeFromMappedType(<MappedType>target) === getIndexType(source)) {

```

Comment:
There's a branch lower down that checks whether `target` is a generic mapped type. Can this be performed there?

Reply:
No because that particular section of code is only entered if `source` is an object or intersection type.